### PR TITLE
Implement FlightSqlProducer - CommandGetCatalogs

### PR DESCRIPTION
**Pull request by rafael-telles**

### CODE REVIEW

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java [hunk 1]
```diff
@@ -67,7 +68,8 @@ public FlightWorkManager(Provider<UserWorker> workerProvider,
    * @return A FlightPreparedStatement which consumes the result of the job.
    */
   public FlightPreparedStatement createPreparedStatement(FlightDescriptor flightDescriptor,
-                                                         Supplier<Boolean> isRequestCancelled, UserSession userSession) {
+                                                         Supplier<Boolean> isRequestCancelled,
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java:71]: Are these formatting changes required? If not, can we avoid them as they clutter the review and make it harder to find what the actual changes were.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java [hunk 2]
```diff
@@ -112,7 +115,20 @@ public void runPreparedStatement(UserProtos.PreparedStatementHandle handle,
     final UserResponseHandler responseHandler = runQueryResponseHandlerFactory.getHandler(runExternalId, userSession,
       workerProvider, optionManagerProvider, listener, allocator);
 
-    workerProvider.get().submitWork(runExternalId, userSession, responseHandler, userRequest, TerminationListenerRegistry.NOOP);
+    workerProvider.get()
+      .submitWork(runExternalId, userSession, responseHandler, userRequest, TerminationListenerRegistry.NOOP);
+  }
+
+  public void getCatalogs(FlightProducer.ServerStreamListener listener, BufferAllocator allocator,
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java:122]: Please add JavaDocs to public methods.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package com.dremio.service.flight.impl;
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:1]: Please ensure all files have a license header.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  public GetCatalogsResponseHandler(BufferAllocator allocator, FlightProducer.ServerStreamListener listener) {
+    this.allocator = allocator;
+    this.listener = listener;
+  }
+
+  @Override
+  public void sendData(RpcOutcomeListener<GeneralRPCProtos.Ack> outcomeListener,
+                       QueryWritableBatch result) {
+  }
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:29]: Is this correct to be empty?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  @Override
+  public void completed(UserResult result) {
+    UserProtos.GetCatalogsResp catalogsResp = result.unwrap(UserProtos.GetCatalogsResp.class);
+
+    try (VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(FlightSqlProducer.Schemas.GET_CATALOGS_SCHEMA,
+      allocator)) {
+      listener.start(vectorSchemaRoot);
+
+      vectorSchemaRoot.allocateNew();
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:39]: This seems wrong - we're sending data on the completed event?

Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:55]: Another implementation of UserResponseHandler:`ResponseSenderHandler`, works just like this and it's used for these catalog functions

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 4 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:45]: Yes - Another implementation of UserResponseHandler:ResponseSenderHandler, works just like this and it's used for these catalog functions

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java (hunk 1 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java]: Sure - we will avoid doing massive formatting

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import com.dremio.exec.work.protector.UserResult;
+
+/**
+ * {@link UserResponseHandler} implementation for {@link FlightWorkManager#getCatalogs}.
+ */
+class GetCatalogsResponseHandler implements UserResponseHandler {
+  private final BufferAllocator allocator;
+  private final FlightProducer.ServerStreamListener listener;
+
+  public GetCatalogsResponseHandler(BufferAllocator allocator, FlightProducer.ServerStreamListener listener) {
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:40]: JavaDoc please.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 5 above)
Comment (kylep-dremio) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:58]: Yes - but see `RunQueryResponseHandler` as what I think is more representative given they both return result sets.

File: services/arrow-flight/src/test/java/com/dremio/service/flight/AbstractTestFlightServer.java [hunk 7]
```diff
@@ -235,4 +235,13 @@ private FlightStream executeQuery(String query) throws IOException, SQLException
       return actualStringResults;
     }
   }
+
+  abstract CallOption[] getCallOptions();
+
+  public FlightInfo getFlightInfo(String query) throws SQLException {
```
Comment (kylep-dremio) [inline services/arrow-flight/src/test/java/com/dremio/service/flight/AbstractTestFlightServer.java:241]: JavaDocs please.

File: services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightServerWithTokenAuth.java [hunk 8]
```diff
@@ -43,10 +44,14 @@ protected String getAuthMode() {
   }
 
   @Override
-  public FlightInfo getFlightInfo(String query) {
+  CallOption[] getCallOptions() {
```
Comment (kylep-dremio) [inline services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightServerWithTokenAuth.java:47]: There are lots of unrelated test changes - I think this is refactoring to get the testing portion ready to properly handle Flight SQL tests of the sort you're now doing?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import com.dremio.exec.proto.GeneralRPCProtos;
+import com.dremio.exec.proto.UserProtos;
+import com.dremio.exec.rpc.RpcOutcomeListener;
+import com.dremio.exec.work.protector.UserResponseHandler;
+import com.dremio.exec.work.protector.UserResult;
+
+/**
+ * {@link UserResponseHandler} implementation for {@link FlightWorkManager#getCatalogs}.
+ */
+class GetCatalogsResponseHandler implements UserResponseHandler {
```
Comment (jduo) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:36]: This class should extend CancellableUserResponseHandler, similar to how we have CreatePreparedStatementResponseHandler.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 4 above)
Comment (jduo) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:48]: Let's throw an invalid operation exception if this method is never called.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    this.listener = listener;
+  }
+
+  @Override
+  public void sendData(RpcOutcomeListener<GeneralRPCProtos.Ack> outcomeListener,
+                       QueryWritableBatch result) {
+  }
+
+  @Override
+  public void completed(UserResult result) {
```
Comment (jduo) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java:51]: Need to handle various responses from result.getState(). This doesn't handle errors.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 4 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java]: Fixed

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 9 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java]: You are right! 
I refactored CancellableUserResponseHandler to handle both cases (CreatePreparedStatement, GetCatalogs and will apply to all others) and deleted the subclasses CreatePreparedStatementResponseHandler and GetCatalogsResponseHandler.

Can you take a look now please? @jduo

File: services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightServerWithTokenAuth.java (hunk 8 above)
Comment (rafael-telles) [reply services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightServerWithTokenAuth.java:47]: Sorry about this @kylep-dremio - yes, this is to properly handle both legacy and FlightSql and both basic and token authentication methods

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 6 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java]: Deleted this class

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java (hunk 10 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/GetCatalogsResponseHandler.java]: Done on CancellableUserResponseHandler

File: services/arrow-flight/src/test/java/com/dremio/service/flight/AbstractTestFlightServer.java (hunk 7 above)
Comment (rafael-telles) [reply services/arrow-flight/src/test/java/com/dremio/service/flight/AbstractTestFlightServer.java:247]: Fixed

File: services/arrow-flight/src/main/java/com/dremio/service/flight/protector/CancellableUserResponseHandler.java [hunk 11]
```diff
@@ -104,4 +149,5 @@ public void cancelJob() {
   protected CompletableFuture<T> getCompletableFuture() {
     return future;
   }
+
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/protector/CancellableUserResponseHandler.java:152]: Nit: extra newline.

File: services/arrow-flight/src/test/java/com/dremio/service/flight/protector/TestCancellableUserResponseHandler.java [hunk 12]
```diff
@@ -141,17 +132,8 @@ private void testClientCancelCaughtAndPropagatedToServer(Supplier<Boolean> isCan
...
-        public void sendData(RpcOutcomeListener<GeneralRPCProtos.Ack> outcomeListener, QueryWritableBatch result) {
-          fail();
-        }
-
-        @Override
-        public void completed(UserResult result) {
-          fail();
-        }
-      };
+      new CancellableUserResponseHandler<>(externalId, userSession, mockedUserWorkerProvider, isCancelled,
```
Comment (jduo) [inline services/arrow-flight/src/test/java/com/dremio/service/flight/protector/TestCancellableUserResponseHandler.java:135]: Doesn't this test no longer fail correctly now that sendData()/completed() calls no longer call fail? Suggest we remove the final marker and leave those overrides in place.

Comment (rafael-telles) [reply services/arrow-flight/src/test/java/com/dremio/service/flight/protector/TestCancellableUserResponseHandler.java]: Not actually, the original test also didn't call sendData() or completed(), what happens if that `CancellableUserResponseHandler#get()` throws the exception that the test is expecting when it gets here: https://github.com/s1mbi0se/dremio-oss/blob/master/services/arrow-flight/src/main/java/com/dremio/service/flight/protector/CancellableUserResponseHandler.java#L83
